### PR TITLE
Martin/ios 384 update flow for waitlist

### PR DESCRIPTION
**Pull request by MartinJibber**

### CODE REVIEW

File: Shared/API/Onboarding+Cloud.swift [hunk 1]
```diff
@@ -76,36 +76,15 @@ struct VerifyCode: CloudFunction {
 struct ActivateUser: CloudFunction {
 
     typealias ReturnType = Any
-    let fullName: String
```
Comment (MartinJibber) [inline Shared/API/Onboarding+Cloud.swift:79]: I'm decoupling activating the user from entering the name. We can save the save the user's name immediately after they enter it.

Let me know if that causes any problems.

File: Shared/Flows/Onboarding/OnboardingCoordinator.swift [hunk 2]
```diff
@@ -43,9 +43,14 @@ extension OnboardingCoordinator: LaunchActivityHandler {
 
 extension OnboardingCoordinator: OnboardingViewControllerDelegate {
     
-    nonisolated func onboardingView(_ controller: OnboardingViewController,
-                                    didVerify user: User) {
+    nonisolated func onboardingViewController(_ controller: OnboardingViewController,
+                                              didOnboard user: User) {
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingCoordinator.swift:47]: This function only gets called after all onboarding is complete.

File: Shared/Flows/Onboarding/OnboardingCoordinator.swift [hunk 3]
```diff
@@ -43,9 +43,14 @@ extension OnboardingCoordinator: LaunchActivityHandler {
...
         Task {
+            await controller.showLoading()
+            try await ActivateUser().makeRequest(andUpdate: [],
+                                                 viewsToIgnore: [controller.view])
+            await controller.hideLoading()
+
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingCoordinator.swift:53]: Once onboarding is complete, we'll activate the user. If the user is on the waitlist, this function won't get called.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 4]
```diff
@@ -70,37 +69,22 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
-//            }
-//        }.store(in: &self.cancellables)
+        self.welcomeVC.onDidComplete = { [unowned self] result in
+            switch result {
+            case .success:
+                self.current = .phone(self.phoneVC)
+            case .failure:
+                break
+            }
+        }
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:79]: We don't have a login button at the moment so just send the user to the phone entry.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 5]
```diff
@@ -70,37 +69,22 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
-//                if let identity = reservation.createdBy?.objectId {
-//                    Task {
-//                        try await self.updateInvitor(with: identity)
-//                    }
-//                }
-//                self.current = .phone(self.phoneVC)
-//            case .reservationError:
-//                self.updateUI()
-//            }
-//        }.store(in: &self.cancellables)
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:95]: I don't think any of this is needed.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 6]
```diff
@@ -112,19 +96,11 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
-                    self.current = .waitlist(self.waitlistVC)
-                    #else
-                    self.showLoading(user: current)
-                    #endif
-                } else {
+
+                switch current.status {
+                case .active:
+                    self.delegate.onboardingViewController(self, didOnboard: current)
+                case .needsVerification, .waitlist, .inactive, .none:
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:103]: When the code entry is complete, we either go to the name field, or finish onboarding if the user is activated.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 7]
```diff
@@ -143,23 +119,17 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
-                Task {
-                    self.showLoading()
-                    guard let fullName = UserDefaultsManager.getString(for: .fullName) else { return }
-                    try await ActivateUser(fullName: fullName).makeRequest(andUpdate: [], viewsToIgnore: [self.view])
-                    await self.hideLoading()
-                    guard let user = User.current(), user.status == .active else { return }
-                    self.delegate.onboardingView(self, didVerify: user)
-                }
+                guard let user = User.current() else { return }
+                self.delegate.onboardingViewController(self, didOnboard: user)
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:132]: Taking a photo is the last step so let the delegate know we're done onboarding.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 8]
```diff
@@ -342,26 +285,52 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
-                User.current()?.formatName(from: name)
-                try await User.current()?.saveLocalThenServer()
-                self.current = .waitlist(self.waitlistVC)
-            }
-            #else
-            if let current = User.current(), current.isOnboarded {
-                self.delegate.onboardingView(self, didVerify: User.current()!)
-            } else {
-                self.current = .photo(self.photoVC)
+    @MainActor
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:288]: Reorganization

File: Shared/Models/Parse Subclasses/User/User.swift [hunk 9]
```diff
@@ -32,10 +32,10 @@ enum UserKey: String {
...
 
 enum UserStatus: String {
-    case active // Has completed onboarding and has full access to the app
-    case waitlist // Has verified phone number and is waiting for access to the app
-    case inactive // Has been given access to the full app but has not completed onboarding
-    case needsVerification // Has entered a phone number but not a verification code. 
+    case needsVerification // Has entered a phone number but not a verification code.
+    case inactive // Has verified but has not completed required onboarding
+    case waitlist // Has verified but has not being granted access to the full app
+    case active // Has verified and fully onboardied. Has full access to the app
```
Comment (MartinJibber) [inline Shared/Models/Parse Subclasses/User/User.swift:38]: Putting these in sequence for my own understanding

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 10]
```diff
@@ -186,93 +149,74 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
...
+                if user.isOnboarded {
+                    self.delegate.onboardingViewController(self, didOnboard: User.current()!)
+                } else {
+                    self.current = .photo(self.photoVC)
+                }
+#endif
+            case .waitlist:
+                self.current = .waitlist(self.waitlistVC)
+            case .active:
+                self.delegate.onboardingViewController(self, didOnboard: user)
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:184]: Now handling all activation scenarios.

File: Shared/API/Onboarding+Cloud.swift (hunk 1 above)
Comment (Benjamindodgson) [reply Shared/API/Onboarding+Cloud.swift:79]: I don't think so, but confirm with @gcolotti

File: Shared/Flows/Onboarding/OnboardingViewController.swift (hunk 5 above)
Comment (Benjamindodgson) [reply Shared/Flows/Onboarding/OnboardingViewController.swift:95]: correct.

File: Shared/Flows/Onboarding/OnboardingViewController.swift (hunk 4 above)
Comment (Benjamindodgson) [reply Shared/Flows/Onboarding/OnboardingViewController.swift:79]: Can you just add one below the waitlist for now? I think we will need it for testing.

File: Shared/Flows/Onboarding/OnboardingViewController.swift [hunk 11]
```diff
@@ -112,19 +96,11 @@ class OnboardingViewController: SwitchableContentViewController<OnboardingConten
                 }
 
                 guard let current = User.current() else { return }
-                if current.isOnboarded, current.status == .active {
-                    #if APPCLIP
-                    self.current = .waitlist(self.waitlistVC)
```
Comment (Benjamindodgson) [inline Shared/Flows/Onboarding/OnboardingViewController.swift:117]: You can delete this waitlist vc

**Review by Benjamindodgson — CHANGES_REQUESTED** [review]
Great work! Just a few things. See comments.

File: Shared/Flows/Onboarding/OnboardingViewController.swift (hunk 4 above)
Comment (MartinJibber) [reply Shared/Flows/Onboarding/OnboardingViewController.swift:79]: @Benjamindodgson Yep, no problem.

**Pull request merged**